src/LanguageServer/Microsoft.CodeAnalysis.LanguageServer/HostWorkspace/LanguageServerProjectLoader.cs

```diff
@@ -273,4 +370,74 @@ async Task LogDiagnosticsAsync(ImmutableArray<DiagnosticLogItem> diagnosticLogIt
             await toastErrorReporter.ReportErrorAsync(worstLspMessageKind, message, cancellationToken);
         }
     }
+
+    protected async ValueTask BeginLoadingProjectAsync(string projectPath, ProjectId primordialProjectId)
+    {
+        using (await _gate.DisposableWaitAsync(CancellationToken.None))
+        {
+            if (_loadedProjects.ContainsKey(projectPath))
+            {
+                return;
+            }
+
+            _loadedProjects.Add(projectPath, new ProjectLoadState.BeginLoadingWithPrimordial(primordialProjectId));
+            _projectsToLoadAndReload.AddWork(new ProjectToLoad(projectPath, ProjectGuid: null, ReportTelemetry: true));
+        }
+    }
+
+    protected async Task LoadProjectsAsync(ImmutableArray<(string ProjectPath, string? ProjectGuid)> projectPaths, CancellationToken cancellationToken)
+    {
+        using (await _gate.DisposableWaitAsync(cancellationToken))
+        {
+            foreach (var (path, guid) in projectPaths)
+            {
+                if (_loadedProjects.ContainsKey(path))
+                {
+                    continue;
+                }
+
+                _loadedProjects.Add(path, new ProjectLoadState.BeginLoading());
+                _projectsToLoadAndReload.AddWork(new ProjectToLoad(Path: path, ProjectGuid: guid, ReportTelemetry: true));
+            }
+
+            // TODO: since this method is used to implement 'OpenSolution' and 'OpenProjects',
+            // it seems like currently loaded projects which are not included in 'projectPaths' should be unloaded.
+        }
+
+        await _projectsToLoadAndReload.WaitUntilCurrentBatchCompletesAsync();
+    }
+
+    protected async ValueTask UnloadProjectAsync(string projectPath)
+    {
+        using (await _gate.DisposableWaitAsync(CancellationToken.None))
+        {
+            if (_loadedProjects.Remove(projectPath, out var loadState))
```

Comment:
can this be a constract call?  i assume it's a bug to try to unload a project not loaded.

Reply:
I think we can't block multiple-unloads without further changes elsewhere. When LspWorkspaceManager finds a document in the HostWorkspace (i.e. the document is contained in an ordinary project), it calls TryRemove for the same Uri on the misc workspace. It pretty much does this over and over and over again as Lsp continues to request new versions of the document.

Reply:
please doc the code :)  thi sis very non-obvious stuff.  and if this is actually what we need to be correct, we should be very explicit on that for future maintainability.  (And potentially test exactly this).  thanks.

Reply:
Added comment.